### PR TITLE
Refactor type visitor walking

### PR DESCRIPTION
**Pull request by oli-obk**
r? @petrochenkov 

pulling out the uncontroversial parts of https://github.com/rust-lang/rust/pull/113671

### CODE REVIEW

File: compiler/rustc_ty_utils/src/sig_types.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        DefKind::Coroutine => {
+            match tcx.type_of(item).instantiate_identity().kind() {
+                ty::Coroutine(_, args, _) => visitor.visit(tcx.def_span(item), args.as_coroutine().sig())?,
+                _ => bug!(),
+            }
+            for (pred, span) in tcx.predicates_of(item).instantiate_identity(tcx) {
+                visitor.visit(span, pred)?;
+            }
+        }
+        // Walk over the signature of the function-like to find the opaques.
```
Comment (petrochenkov) [inline compiler/rustc_ty_utils/src/sig_types.rs:37]: Outdated comment, this function is no longer just about opaques.
(And one more similar comment below.)

**Comment by petrochenkov** [comment]
r=me with the comments updated.
@rustbot author

**Comment by xxchan** [comment]
Hi, this PR caused a regression for us. Don't know if it's expected. :eyes: 

---

searched nightlies: from nightly-2023-10-25 to nightly-2023-10-27
regressed nightly: nightly-2023-10-26
searched commit range: https://github.com/rust-lang/rust/compare/df871fbf053de3a855398964cd05fadbe91cf4fd...ab5c841a1f3c09edc5ea07722519627c960aed17
regressed commit: https://github.com/rust-lang/rust/commit/c2ef35161fc7477b38f2e556be2fd6d85d9f4905

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>


Host triple: aarch64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start 2023-10-25 --end 2023-10-27 -- check -p risingwave_storage
```
</details>

---

There's a lot of  "non-defining opaque type use in defining scope" errors. 
https://github.com/risingwavelabs/risingwave/actions/runs/6673382874/job/18138864486 (Sorry for not minimized)
```
error[E0792]: non-defining opaque type use in defining scope
   --> src/storage/src/mem_table.rs:444:9
    |
444 | /         async move {
445 | |             let stream = self
446 | |                 .inner
447 | |                 .iter(key_range.clone(), self.epoch(), read_options)
...   |
454 | |             ))
455 | |         }
    | |_________^ argument `'_` is not a generic parameter
    |
note: for this opaque type
   --> src/storage/src/mem_table.rs:304:1
    |
304 | #[try_stream(ok = StateStoreIterItem, error = StorageError)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

**Comment by oli-obk** [comment]
Nope, no regression was expected. I'll try to find a minimization on Monday. You're using a lot of unstable features and proc macros, so i'll try to come up with an example by looking at this PRs diff instead of minimizing your crate

**Comment by oli-obk** [comment]
Not minimal, I'll run it through a minimizer after lunch

```rust
// edition: 2021

#![feature(impl_trait_in_assoc_type)]

use std::future::Future;

pub struct MemtableLocalStateStore<S: StateStoreWrite + StateStoreRead> {
    mem_table: MemTable,
    inner: S,

    epoch: u64,

    table_id: TableId,
    is_consistent_op: bool,
    table_option: TableOption,
}

impl<S: StateStoreWrite + StateStoreRead> LocalStateStore for MemtableLocalStateStore<S> {
    type IterStream<'a> = impl StateStoreIterItemStream + 'a where Self: 'a;

    fn iter(
        &self,
        key_range: TableKeyRange,
        read_options: ReadOptions,
    ) -> impl Future<Output = StorageResult<Self::IterStream<'_>>> + '_ {
        async move {
            let stream = self.inner.iter(key_range.clone(), self.epoch, read_options).await?;
            Ok(merge_stream(self.mem_table.iter(key_range), stream, self.table_id, self.epoch))
        }
    }
}

trait StateStoreRead {
    async fn iter(
        &self,
        _: TableKeyRange,
        _: u64,
        _: ReadOptions,
    ) -> StorageResult<impl StateStoreReadIterStream>;
}

trait StateStoreWrite {}
trait LocalStateStore {
    type IterStream<'a>
    where
        Self: 'a;

    fn iter(
        &self,
        key_range: TableKeyRange,
        read_options: ReadOptions,
    ) -> impl Future<Output = StorageResult<Self::IterStream<'_>>> + '_;
}
trait StateStoreIterItemStream {}

impl<T> StateStoreIterItemStream for T {}

struct MemTable;
struct TableOption;

impl MemTable {
    fn iter<'a>(
        &'a self,
        _: TableKeyRange,
    ) -> impl Iterator<Item = (&'a TableKey<Bytes>, &'a KeyOp)> + 'a {
        std::iter::empty()
    }
}

trait StateStoreReadIterStream {}

#[derive(Copy, Clone)]
struct TableId;
struct ReadOptions;
#[derive(Copy, Clone)]
struct TableKeyRange;

pub(crate) async fn merge_stream<'a>(
    mem_table_iter: impl Iterator<Item = (&'a TableKey<Bytes>, &'a KeyOp)> + 'a,
    inner_stream: impl StateStoreReadIterStream,
    table_id: TableId,
    epoch: u64,
) {
}

type StorageResult<T> = Result<T, ()>;

struct TableKey<T>(T);

struct Bytes;

struct KeyOp;

fn main() {}
```

**Comment by oli-obk** [comment]
https://github.com/rust-lang/rust/pull/117371 silences the `non-defining opaque type use` error on the `iter` return type. the `merge_stream` return type error persists

**Pull request merged**